clash-prelude/tests/Clash/Tests/Laws/SaturatingNum.hs

```diff
@@ -0,0 +1,195 @@
+  , (5,  pure 0)
+  , (85, Gen.integral (Range.linear minBound maxBound)) ]
+
+genIndex :: forall n. KnownNat n => Gen (Index n)
+genIndex = genBoundedIntegral
+
+genUnsigned :: forall n. KnownNat n => Gen (Unsigned n)
+genUnsigned = genBoundedIntegral
+
+genSigned :: forall n. KnownNat n => Gen (Signed n)
+genSigned = genBoundedIntegral
+
+-- | Generates a bounded fractional with a bias towards extreme values:
+--
+--      5%: minBound
+--      5%: maxBound
+--      5%: 0.0
+--     85%: uniform [minBound, maxBound]
+--
+genBoundedFractional :: forall a. (Real a, Fractional a, Bounded a) => Gen a
+genBoundedFractional = Gen.frequency
+  [ (5,  pure minBound)
+  , (5,  pure maxBound)
+  , (5,  pure 0.0)
+  , (85,   fmap (fromRational . toRational)
+         $ Gen.double
+         $ fmap fromRational
+         $ Range.linearFrac
+             (toRational (minBound @a))
+             (toRational (maxBound @a))) ]
+
+genSFixed :: forall a b. (KnownNat a, KnownNat b) => Gen (SFixed a b)
+genSFixed = genBoundedFractional
+
+genUFixed :: forall a b. (KnownNat a, KnownNat b) => Gen (UFixed a b)
+genUFixed = genBoundedFractional
+
+tests :: TestTree
+tests = testGroup "SaturatingNum"
+  [ -- testSaturationLaws True "Index 1" (genIndex @1)
+    testSaturationLaws True "Index 2" (genIndex @2)
+  , testSaturationLaws True "Index 128" (genIndex @128)
+
+  , testSaturationLaws True "Unsigned 0" (genUnsigned @0)
+  , testSaturationLaws True "Unsigned 1" (genUnsigned @1)
+  , testSaturationLaws True "Unsigned 32" (genUnsigned @32)
+  , testSaturationLaws True "Unsigned 127" (genUnsigned @127)
+  , testSaturationLaws True "Unsigned 128" (genUnsigned @128)
+
+  -- , testSaturationLaws True "Signed 0" (genSigned @0)
+  -- , testSaturationLaws True "Signed 1" (genSigned @1)
+  , testSaturationLaws True "Signed 32" (genSigned @32)
+  , testSaturationLaws True "Signed 127" (genSigned @127)
+  , testSaturationLaws True "Signed 128" (genSigned @128)
+
+  -- , testSaturationLaws False "SFixed 0 0" (genSFixed @0 @0)
+  -- , testSaturationLaws False "SFixed 0 1" (genSFixed @0 @1)
+  -- , testSaturationLaws False "SFixed 1 0" (genSFixed @1 @0)
+  -- , testSaturationLaws False "SFixed 1 1" (genSFixed @1 @1)
+  -- , testSaturationLaws False "SFixed 1 2" (genSFixed @1 @2)
+  -- , testSaturationLaws False "SFixed 2 1" (genSFixed @2 @1)
+  -- , testSaturationLaws False "SFixed 2 2" (genSFixed @2 @2)
+  -- , testSaturationLaws False "SFixed 128 128" (genSFixed @128 @128)
+
+  , testSaturationLaws False "UFixed 0 0" (genUFixed @0 @0)
+  , testSaturationLaws False "UFixed 0 1" (genUFixed @0 @1)
+  , testSaturationLaws False "UFixed 1 0" (genUFixed @1 @0)
+  , testSaturationLaws False "UFixed 1 1" (genUFixed @1 @1)
+  , testSaturationLaws False "UFixed 1 2" (genUFixed @1 @2)
+  , testSaturationLaws False "UFixed 2 1" (genUFixed @2 @1)
+  , testSaturationLaws False "UFixed 2 2" (genUFixed @2 @2)
+  , testSaturationLaws False "UFixed 128 128" (genUFixed @128 @128)
```

Comment:
These should all fail when PR #1492 is merged, because `satSucc SatWrap maxBound  /= minBound` for `Fixed`! Similary for `satPred`.

Reply:
Great. Let's get your PR in first then, I'll see what I can do after.